pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -799,7 +856,22 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
 			}
 		}
 	}
-	p.moveRequestCycle = p.schedulingCycle
+
+	// Besides setting MoveRequestCycle above for all pods which got moved,
+	// we also need to set it for pods that are being processed because
+	// AddUnschedulableIfNotPresent might get called for them, and in
+	// AddUnschedulableIfNotPresent we need to know whether events were
+	// observed while scheduling them.
+	for uid, inFlight := range p.inFlightPods {
+		logger.V(5).Info("In-flight pod with concurrent event", "pod", klog.KObj(inFlight.pod), "event", event.Label, "schedulingCycle", p.schedulingCycle)
+
+		inFlight.events = append(inFlight.events, clusterEvent{
+			event:        event,
+			requestCycle: p.schedulingCycle,
+		})
```

Comment:
> the number of inflighPods can be potentially unbounded.

This is a very good point.. I checked how the number of goroutines for the binding cycle will increase in a very simple case ([SchedulingBasic/5000Nodes](https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler_perf/config/performance-config.yaml#L18))

- if everything is usual -> at max around 10 binding cycles running at the same time 
- if all binding cycle takes 1 second to finish (e.g., kube-apiserver is temporary unstable and slower than usual) -> at max around 600 binding cycles running at the same time 🤯

What should we move events from each inFlight to somewhere else somehow so that we don't need to have events in every inFlight Pods? Or do you have any other ideas here?

Reply:
Oh wow, that's a lot of inflight requests.

At the high level, I can see two approaches:
- Limit the number of inflight binding operations, this should be possible by using a buffered channel, instead of directly calling the routine.
- Have a single list of events and then each pod has a pointer to the position in the list when they entered the inflight set. A single linked list might do the trick.

@tosi3k do you have any advice from a scalability perspective?
This is related to the inflight binding operations (set nodeName) after kube-scheduler has made a decision.

Reply:
Alright, I'll try to do the second in this weekend. 
Your first one might be needed in another context, but it won't reduce the number of in-flight Pods, right? That would contribute to reducing the number of goroutines of binding but not reducing the number of in-flight Pods waiting for binding cycle. then the queue needs to have those inflight Pods until they're bound (or failed to be bound and moved back to the queue)

Reply:
>  That would contribute to reducing the number of goroutines of binding but not reducing the number of in-flight Pods waiting for binding cycle

It would, because the scheduling cycle would be blocked waiting for the buffered channel to have space.

Reply:
> the scheduling cycle would be blocked waiting for the buffered channel to have space.

Ah, so you mean the scheduler wouldn't start the next scheduling cycle until the number of goroutines for the binding cycle gets reduced. That makes sense.

Reply:
I've already got [the first one](https://github.com/kubernetes/kubernetes/pull/118438#discussion_r1232415391) in this PR. Do you expect to limit the number of the binding cycle to be implemented in this PR as well?